### PR TITLE
Add test fixture for experimental-lto-mode

### LINKED ISSUE #6888: Can not build swift-nio example `NIOUDPEchoServer` with `--experimental-lto-mode=full`
### Description

I tried to build the `NIOUDPEchoServer` with lto enabled.

### Expected behavior

`NIOUDPEchoServer` can be build in release mode with full lto optimization.

### Actual behavior

```
[root@10f889c21a1b src]# swift run -c release --static-swift-stdlib --experimental-lto-mode=full NIOUDPEchoServer
Building for production...
error: autolink-extract command failed with exit code 1 (use -v to see invocation)
<unknown>:0: error: error opening input file '/src/.build/aarch64-unknown-linux-gnu/release/CNIOAtomics.build/src/c-atomics.c.o' (The file was not recognized as a valid object file
)
error: fatalError
[1/2] Linking NIOUDPEchoServer
```

### Steps to reproduce

1. Checkout NIO (`git clone --depth 1 https://github.com/apple/swift-nio.git`)
2. Run inside the package `swift run -c release --static-swift-stdlib --experimental-lto-mode=full NIOUDPEchoServer`

### Swift Package Manager version/commit hash

_No response_

### Swift & OS version (output of `swift --version ; uname -a`)

```
[root@10f889c21a1b src]# swift --version ; uname -a
Swift version 5.9-dev (LLVM 50c867d95965de6, Swift ca3e3aa237df721)
Target: aarch64-unknown-linux-gnu
Linux 10f889c21a1b 5.15.111 #2 SMP Mon May 15 16:54:21 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux
```

### PR DESCRIPTION
**Pull request by rauhul**
Adds a test covering experimental lto mode and fixes a bug where the
lto mode was not passed to the linker.

Fixes #6888

### CODE REVIEW

**Comment by rauhul** [comment]
@swift-ci please smoke test

**Comment by rauhul** [comment]
@swift-ci please test

**Comment by rauhul** [comment]
```console
[12/14] Write Objects.LinkFileList
error: autolink-extract command failed with exit code 1 (use -v to see invocation)
<unknown>:0: error: error opening input file '/tmp/Miscellaneous_LTO_SwiftAndCTargets.GHkhKS/Miscellaneous_LTO_SwiftAndCTargets/.build/x86_64-unknown-linux-gnu/debug/cLib.build/cLib.c.o' (The file was not recognized as a valid object file)
```

@fabianfett it looks lto (as currently implemented) just doesn't work on linux! Thanks for the report!

**Comment by neonichu** [comment]
Maybe @artemcm does

**Comment by artemcm** [comment]
`swift-autolink-extract` takes as input all the object files from compilation commands in order to extract auto-linking directives so that they can be specified as explicit flags to the linker, because non-Darwin linkers do not have auto-linking feature. It is strange why LLVM is not recognizing this object file as a valid `.o`.

**Comment by rauhul** [comment]
> `swift-autolink-extract` takes as input all the object files from compilation commands in order to extract auto-linking directives so that they can be specified as explicit flags to the linker, because non-Darwin linkers do not have auto-linking feature. It is strange why LLVM is not recognizing this object file as a valid `.o`.

Well this file doesn't exist on disk because the file extension is `.bc` right?

**Comment by artemcm** [comment]
> > `swift-autolink-extract` takes as input all the object files from compilation commands in order to extract auto-linking directives so that they can be specified as explicit flags to the linker, because non-Darwin linkers do not have auto-linking feature. It is strange why LLVM is not recognizing this object file as a valid `.o`.
> 
> Well this file doesn't exist on disk because the file extension is `.bc` right?
> 
> ```shell
> ./.build/aarch64-unknown-linux-gnu/debug/swiftLib.build/swiftLib.swift.bc
> ./.build/aarch64-unknown-linux-gnu/debug/exe.build/main.swift.bc
> ```

Ah, then we need to teach the driver about how this works, I believe it is just forwarding compilation outputs to `swift-autolink-extract`.

**Comment by rauhul** [comment]
> > > `swift-autolink-extract` takes as input all the object files from compilation commands in order to extract auto-linking directives so that they can be specified as explicit flags to the linker, because non-Darwin linkers do not have auto-linking feature. It is strange why LLVM is not recognizing this object file as a valid `.o`.
> > 
> > 
> > Well this file doesn't exist on disk because the file extension is `.bc` right?
> > ```shell
> > ./.build/aarch64-unknown-linux-gnu/debug/swiftLib.build/swiftLib.swift.bc
> > ./.build/aarch64-unknown-linux-gnu/debug/exe.build/main.swift.bc
> > ```
> 
> Ah, then we need to teach the driver about how this works, I believe it is just forwarding compilation outputs to `swift-autolink-extract`.

I'm a little confused because we use `.bc` because that's the extension the driver chooses

**Comment by artemcm** [comment]
> > > > `swift-autolink-extract` takes as input all the object files from compilation commands in order to extract auto-linking directives so that they can be specified as explicit flags to the linker, because non-Darwin linkers do not have auto-linking feature. It is strange why LLVM is not recognizing this object file as a valid `.o`.
> > > 
> > > 
> > > Well this file doesn't exist on disk because the file extension is `.bc` right?
> > > ```shell
> > > ./.build/aarch64-unknown-linux-gnu/debug/swiftLib.build/swiftLib.swift.bc
> > > ./.build/aarch64-unknown-linux-gnu/debug/exe.build/main.swift.bc
> > > ```
> > 
> > 
> > Ah, then we need to teach the driver about how this works, I believe it is just forwarding compilation outputs to `swift-autolink-extract`.
> 
> I'm a little confused because we use `.bc` because that's the extension the driver chooses

This job must not be configured correctly. It cannot accept `.bc` files, so it must not have been taught about these compilation flows before now.

**Comment by rauhul** [comment]
> This job must not be configured correctly. It cannot accept `.bc` files, so it must not have been taught about these compilation flows before now.

Is this an spm shouldn't be emitting this job or swift-driver?

**Comment by rauhul** [comment]
@artemcm would teaching `lib/DriverTool/autolink_extract_main.cpp` about llvm bitcode files (how to open/load/search for the auto link section) be the right approach to fix this?

**Comment by artemcm** [comment]
> @artemcm would teaching `lib/DriverTool/autolink_extract_main.cpp` about llvm bitcode files (how to open/load/search for the auto link section) be the right approach to fix this?

Reading the code, I get the impression that if we're using LTO, then we're using `lld`, which natively supports auto-linking with ELF. 
(https://github.com/apple/swift/blob/3a05e798e6d045bb25e7539e07b870994bc88884/lib/IRGen/IRGenModule.cpp#L1701)

So I wonder if we can simply teach the driver that this flow does not require `autolink-extract`, and the rest of compilation downstream will "just work". Are you able to check this at desk? I won't be near a Linux box until tomorrow.

**Comment by artemcm** [comment]
Hmm, the driver is already supposed to skip `swift-autolink-extract` when it knows we're doing LTO. 
https://github.com/apple/swift/blob/3a05e798e6d045bb25e7539e07b870994bc88884/lib/Driver/Driver.cpp#L2255
https://github.com/apple/swift-driver/blob/e48171a466d624abdeaee3ea5d7a8b7237fda1eb/Sources/SwiftDriver/Jobs/AutolinkExtractJob.swift#L21

So we need to figure out why we're still generating this job here.

**Comment by rauhul** [comment]
> Hmm, the driver is already supposed to skip `swift-autolink-extract` when it knows we're doing LTO. https://github.com/apple/swift/blob/3a05e798e6d045bb25e7539e07b870994bc88884/lib/Driver/Driver.cpp#L2255 https://github.com/apple/swift-driver/blob/e48171a466d624abdeaee3ea5d7a8b7237fda1eb/Sources/SwiftDriver/Jobs/AutolinkExtractJob.swift#L21
> 
> So we need to figure out why we're still generating this job here.

Oh I see, that is interesting.

> Are you able to check this at desk? I won't be near a Linux box until tomorrow.

I've been running under docker to repro this issue. But I guess the question is why the driver doesn't see this link command as an lto job. 

Lemme try to get the full swiftc command and maybe that will provide some clues.

**Comment by rauhul** [comment]
@artemcm here's a verbose build log:
[build.log](https://github.com/apple/swift-package-manager/files/12653370/build.log)

**Comment by rauhul** [comment]
Maybe its as simple as adding `--lto <#LTOKind#>` to the swiftc call which spm produces

**Comment by rauhul** [comment]
@swift-ci please test

**Comment by rauhul** [comment]
@swift-ci please smoke test windows

**Comment by rauhul** [comment]
@swift-ci please test

**Comment by rauhul** [comment]
@swift-ci please smoke test windows

**Comment by rauhul** [comment]
@swift-ci please test

**Comment by rauhul** [comment]
@swift-ci please smoke test windows

**Comment by shahmishal** [comment]
@swift-ci please smoke test windows

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by rauhul** [comment]
@swift-ci please smoke test linux

**Comment by rauhul** [comment]
@swift-ci please smoke test

**Comment by rauhul** [comment]
@swift-ci please test

**Comment by rauhul** [comment]
https://github.com/apple/swift/pull/69696
@swift-ci please test

**Comment by rauhul** [comment]
https://github.com/apple/swift/pull/69564
@swift-ci please test

**Comment by rauhul** [comment]
@swift-ci please test

**Comment by rauhul** [comment]
@swift-ci please test

**Comment by rauhul** [comment]
@swift-ci please test

File: Tests/FunctionalTests/MiscellaneousTests.swift [hunk 1]
```diff
@@ -337,6 +337,31 @@ class MiscellaneousTestCase: XCTestCase {
...
+        // Currently swift-ci does not build/install swift-driver before running
+        // swift-package-manager tests which results in this test failing.
+        // See the following additional discussion:
+        // - https://github.com/apple/swift/pull/69696
+        // - https://github.com/apple/swift/pull/61766
+        // - https://github.com/apple/swift-package-manager/pull/5842#issuecomment-1301632685
+        try fixture(name: "Miscellaneous/LTO/SwiftAndCTargets") { fixturePath in
+            do {
+                let output = try executeSwiftBuild(
+                  fixturePath,
```
Comment (MaxDesiatov) [inline Tests/FunctionalTests/MiscellaneousTests.swift:352]: This seems to be using a mix of 4 and 2-space formatting. Additionally, why wrap this in `do`/`catch` if `try fixture(name:)` would be rethrowing the error anyway?

Comment (rauhul) [reply Tests/FunctionalTests/MiscellaneousTests.swift:352]: I had added `XCTFail("\(error)")` to help with debugging, removing and fixing indentation

**Comment by rauhul** [comment]
@swift-ci please test

**Review by MaxDesiatov — APPROVED** [review]
Thanks!

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Pull request merged**